x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java

```diff
@@ -386,15 +424,38 @@ private static int bitSetLength(int totalValues) {
         return (totalValues + 7) / 8;
     }
 
-    private static void accumulateVectorValidity(List<ArrowBuf> bufs, List<BufWriter> bufWriters, Block b) {
-        bufs.add(dummyArrowBuf(bitSetLength(b.getPositionCount())));
+    /**
+     * Get the value count for a block. For single-valued blocks this is the same as the position count.
+     * For multivalued blocks, this is the flattened number of items.
+     */
+    static int valueCount(Block block) {
+        int result = block.getFirstValueIndex(block.getPositionCount());
+
+        // firstValueIndex is always zero for all-null blocks.
+        if (result == 0 && block.areAllValuesNull()) {
+            result = block.getPositionCount();
+        }
+
+        return result;
+    }
+
+    private static void accumulateVectorValidity(List<ArrowBuf> bufs, List<BufWriter> bufWriters, Block b, boolean multivalued) {
+        // If that block is in a multivalued-column, validities are output in the parent Arrow List buffer (values themselves
+        // do not contain nulls per docvalues limitations).
```

Comment:
Does that mean that for a block like `[1, 2], null, null, [3, 4]` the validity vector of the _offset list_ will be `1001` (two nulls in the middle)?

Reply:
Yes, exactly. Since Arrow supports null values in multivalued columns, there is a validity vector for the offset list and one for the value list. Validity for the offset list is taken care of in `addListOffsets`.